### ISSUE TITLE
Enhancement: [no-throw-literal] option to also check promise rejections

### ISSUE THREAD

**Issue by DetachHead** [typescript-eslint/typescript-eslint#7673]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-throw-literal

### Description

a promise rejected with a non-`Error` value has the same problems as a `throw` statement with a non-`Error` value, so it should also check promise rejections

### Fail

```typescript
Promise.reject('asdf')
new Promise((_, reject) => reject('asdf'))
```

### Pass

```typescript
Promise.reject(new Error('asdf'))
new Promise((_, reject) => reject(new Error('asdf')))
```

**Comment by bradzacher** [comment]
Thanks for filing!
This is something that should be proposed upstream as this is an extension rule.
We don't look to break from the checks that the base rule does - only enrich them with type information.

Basic checks for this would not require type information - promise constructors are easy to handle syntactically and they have rules that do this already (eg [`no-async-promise-executor`](https://eslint.org/docs/latest/rules/no-async-promise-executor))

If they accept it and it is implemented - then we can similarly implement it along with the type-aware enhancements. If they do not accept it - then we shan't either.

I will close this for now. If it is accepted upstream before the auto-lock window, feel free to comment and we can reopen and mark it as blocked upstream. When it is implemented we can mark it as accepting PRs.

**Comment by DetachHead** [comment]
raised https://github.com/eslint/eslint/issues/17592

**Comment by DetachHead** [comment]
@bradzacher turns out there's already an upstream rule for it https://eslint.org/docs/latest/rules/prefer-promise-reject-errors

can we re-open this issue or would you rather i raise a new one?

**Comment by DetachHead** [comment]
raised separate issue: #7687
